### PR TITLE
Batched Hessian approximation via emp. Fisher and GGN

### PR DESCRIPTION
**Pull request by severinbratus**
- Two modes of operation possible: unbatched (the old way, pass a zip of x's and y's), unbatched (the new way, pass a MLUtils.DataLoader)
- Batched gradients of the loss function (which is a Jacobian, transposed)
- Batched Jacobians of the NN output (which is a Jacobian, reshaped)
- Tested with multiple batchsizes (1, 32)
- Benchmarked (see notebooks)
- Removed `permutedims` call from Jacobians (which is equivalent to transpose), to make Jacobians consistent with their definition (the shape of a Jacobian of f(x) in R^m wrt x in R^n is m by n)
- Removed row-wise linearization of theta (undoing some changes in #17 and #20), and verified that the order does not affect the predictions, as long as it is consistently followed across the codebase. This, however makes our Hessians not directly comparable to those of laplace-torch -- which means that to implement point 2 of #25, we would require utility  reordering functions to transform one Hessian or gradient form to the other.
- Refactored the workflow tests to run on all combinations of backend, batchsize, and subset_of_weights, which multiplies the time it takes to run tests by a factor of `2 * 3 * 3 == 18`. This may be an overkill. Also with some refactoring it should be possible to train a NN only once for a task type.

New deps:
- MLUtils.jl (part of Flux), for DataLoader
- Tullio.jl, for optimized batched einsum computation
- Compat.jl, for compatibility with versions < 1.9 (function `stack` was introduced in 1.9)

Solves #3, #4, for both emp. Fisher and GGN.

Meta TODO: divide into smaller issues

### CODE REVIEW

**Comment by severinbratus** [comment]
Note: the `eeeerw` user is me, before configuring my email properly

**Review by adelinacazacu — APPROVED** [review]
This is a well documented, clean implementation. I haven't been able to run it and get exceptional results on my machine, since it is configured to use the GPU to parallelize computations, but I only have the good old Intel UHD Graphics. :))
However, I see the good results batching gives in the notebooks when you ran it on your machine. The only thing I could nitpick about this PR would be in regards to the test suite, but we agreed previously that we will create separate PRs for improving tests later in the project. All good by me, ready to be merged!

File: dev/notebooks/batching/gif-01.gif
Comment (severinbratus) [inline dev/notebooks/batching/gif-01.gif:1]: For clarity, this is how functional batching *should not* look like.

File: dev/notebooks/batching/Multi-class_classification.jl.ipynb
Comment (MarkArdman) [inline dev/notebooks/batching/Multi-class_classification.jl.ipynb:1]: When trying to run notebooks in the respective subfolder it errors out, I believe it's because we are trying to activate the folder they are in rather than LaplaceRedux

Comment (MarkArdman) [inline dev/notebooks/batching/Multi-class_classification.jl.ipynb:1]: I believe that if we want anyone to be able to run the notebooks as-is, instead of doing ```Pkg.activate(".")``` we should instead do ```Pkg.activate("../../..")```.

Comment (severinbratus) [reply dev/notebooks/batching/Multi-class_classification.jl.ipynb:1]: What is the reasoning behind this? I have installed LaplaceRedux with `]dev /home/.../LaplaceRedux.jl/` (+ I run Revise.jl, not sure if that matters), so the package I load with `using` is the development version.

Comment (MarkArdman) [reply dev/notebooks/batching/Multi-class_classification.jl.ipynb:1]: Indeed that's not a problem as long as u do `]dev`. I was thinking more for people who don't have the development version in the julia registry, but I suppose that doesn't really matter

**Review by MarkArdman — APPROVED** [review]
The implemented changes are very well documented. Everything works exactly as expected, every step of the way is dissected and benchmarked in the notebooks. I quite the addition of helper functions for testing. Overall great changes, maybe some extra comments would do good but they are hardly necessary.
